### ISSUE TITLE
[RFC] type mismatch error message improvement

### ISSUE THREAD

**Issue by timotheecour** [nim-lang/Nim#8200]
The following would be clearer and more concise; especially useful when lots of overloads are shown:

```
test.nim(883, 12) Error: type mismatch: got <Stream, TaintedString>
but expected one of:
proc readLine(f: File; line: var TaintedString): bool
  first type mismatch at position: 1
  required type: File
  but expression 'outp' is of type: Stream
proc readLine(f: File): TaintedString
  first type mismatch at position: 1
  required type: File
  but expression 'outp' is of type: Stream

expression: readLine(outp, line)
      if outp.readLine(line):
```
=>
```
test.nim(883, 12) {if outp.readLine(line):} Error: type mismatch: got <Stream, TaintedString> in expression: readLine(outp, line); candidates:
  proc readLine(f: File {Stream}; line: var TaintedString): bool
  proc readLine(f: File {Stream}): TaintedString
```

**Comment by dom96** [comment]
The compiler should use colours and arrows for this:

```
if outp.readLine(line):
     ^----- `outp` is of type 'Stream' which doesn't match any candidates
test.nim(883, 12) Error: type mismatch: got <Stream, TaintedString>
but expected one of:
proc readLine(*r*f: File*r*; *g*line: var TaintedString*g*): bool
proc readLine(*r*f: File*r*): TaintedString
```

Where text delimited by ``*r*`` is red and by ``*g*`` is green.

**Comment by andreaferretti** [comment]
I actually like the existing version. It is a little verbose, but for a good reason: for each candidate it shows exactly why the matching didn't work. All other versions are not only shorter, but less explicit. There were many issues requesting better error messages along the lines of those of - say - Elm, and this one requests to go back to a less explicit form

**Comment by Vindaar** [comment]
I agree with @andreaferretti. Although adding a little bit of color would be a great addition!

**Comment by data-man** [comment]
> It is a little verbose, but for a good reason: for each candidate it shows exactly why the matching didn't work. 

What if the verbosed version will shown only with ```--verbosity:2 (or 3)``` switch?

**Comment by andreaferretti** [comment]
@data-man I think the verbose version is most useful for beginners, which probably do not even know about the verbosity switch. Maybe `--shortErrorMessages`?

**Comment by timotheecour** [comment]
@cheatfate 
> @dom96 variant is much better, because it actually shows and highlights position of wrong argument, but your variant @timotheecour is hiding position at all.

how is it hiding it?
```
proc readLine(f: File {Stream}; line: var TaintedString): bool
proc readLine(f: File {Stream}): TaintedString
```
shows the first argument is wrong, showing the mismatch between expected (File) and actual (Stream).

Obviously, colors can be added.

**Comment by cheatfate** [comment]
@timotheecour, your example is using just 2 arguments, try to use like 4-6 arguments and like 10-20 variants. Original version shows number of argument, while your version exactly hides this information and if function has much more arguments, your error output become much less valuable to quickly find a problem.

**Comment by timotheecour** [comment]
@cheatfate here's an updated version which shows number of arguments when there's a number of argument mismatch, and shows `A {B}` when there's no number of argument mismatch but a type mismatch:
```
test.nim(883, 12) {if outp.readLine(line, 1, 1+1):} Error: type mismatch: got <Stream, TaintedString, int, int> in expression: readLine(outp, line, 1, 1+1); candidates:
  proc readLine(f: File {Stream}; line: var TaintedString, a: A, b: B): bool
  proc readLine(f: Stream; line: var string {TaintedString}, a: A, b: B): bool
  proc readLine(f: Stream; line: var string, a: A): bool (3 arguments expected, got 4)
  proc readLine(f: Stream; line: var string): bool (2 arguments expected, got 4)
```

**Comment by andreaferretti** [comment]
@timotheecour The existing version pinpoints exactly, for each variant, which arguments do not match. For your version, once has to manually figure out by comparing the expected arguments with all the variants. This is why I find the current version more valuable

**Comment by alehander42** [comment]
What's the status here? That would be a very nice error improvement. 
I really like  @timotheecour 's format + @dom96 's color hints. The only thing I'd add would be maybe
not repeating proc everywhere like in https://github.com/nim-lang/Nim/issues/8656

```
proc readLine:
  (args)
  (args)
```
@andreaferretti I absolutely don't agree, something like @dom96's `^` or color hints keeps the information there without taking a screen of space.

**Comment by andreaferretti** [comment]
I find the lines like

```
  first type mismatch at position: 1
  required type: File
  but expression 'outp' is of type: Stream
```

valuable, and they are lost in the other version, color or not

**Comment by alehander42** [comment]
but the info is not lost: with color or `^` you can see which argument doesn't match and then you can see the original argument there. 

you can even have `required: File, got: Stream` in the end of the line

however having 3 lines for this , when it can be expressed in much less is a waste of space and not nice with long lists

**Comment by andreaferretti** [comment]
@alehander42 Yes, in the code snippet in #8656 the short version is a little more clear. But in the most common case, where you have few overloads and more arguments, I find it easier to pinpoint which argument does not match in each overload, at the cost of being a little repetitive

**Comment by Araq** [comment]
I'm sure we can come up with a solution that is just as clear but a little less verbose.

**Comment by andreaferretti** [comment]
@alehander42 I am not sure I understand exactly what you describe, but I guess that would do in this case. The issue is that different overloads may have different combinations of parameters that match. To make a concrete example, say I write

```nim
let
  s = "hi there"
  a = @[1, 2, 3]
discard(a & s)
```

The current error reads

```
Error: type mismatch: got <seq[int], string>
but expected one of: 
proc `&`[T](x, y: seq[T]): seq[T]
  first type mismatch at position: 2
  required type: seq[T]
  but expression 's' is of type: string
proc `&`[T](x: T; y: seq[T]): seq[T]
  first type mismatch at position: 2
  required type: seq[T]
  but expression 's' is of type: string
proc `&`[T](x: seq[T]; y: T): seq[T]
  first type mismatch at position: 2
  required type: T
  but expression 's' is of type: string

expression: a & s
```

To me, this is missing a few other overloads (I think this is just a bug), so a complete error would also include

```
proc `&`(x: string; y: char): string
  first type mismatch at position: 1
  required type: string
  but expression 'a' is of type: seq[int]
proc `&`(x, y: string): string
  first type mismatch at position: 1
  required type: string
  but expression 'a' is of type: seq[int]
```

In total, we have 5 overloads, and sometimes the mismatch is at position 1, sometimes at position 2, depending on the overload.

Let me try to write this in your suggested style

```
discard(a & s):
            ^----- `s` is of type 'string' which doesn't match any candidates
test.nim(883, 12) Error: type mismatch: got <seq[int], string>
but expected one of:
proc `&`[T](x, y: seq[T]): seq[T]
proc `&`[T](x: T; y: seq[T]): seq[T]
proc `&`[T](x: seq[T]; y: T): seq[T]
proc `&`(x: string; y: char): string
proc `&`(x, y: string): string
```

This is (may be) misleading: possibly the error is actually in position 1, so `s` is correct but `a` is not. The compiler cannot show with an arrow where the error is, because we are not certain - both arguments could be the culprit.

Using colors to show matches/mismatches on the other hand may be a good way to go, because then we have a different information in each overload. The user still has to compare this with the provided arguments, but it is an ok solution

**Comment by Araq** [comment]
> To me, this is missing a few other overloads (I think this is just a bug), so a complete error would also include

No, that was a deliberate change by me. The idea is that we treat the procs as attached to the first parameter's type and so cut these out. That's pretty much how C# does it and C# is doing fine.

**Comment by andreaferretti** [comment]
Then most of what I am saying makes little sense. My whole point was that in the current version you get the full list of overloads and you can check which one has mismatches and in which position.

Since in Nim proc are **not** actually attached to the first parameter - which has no special relevance in the language, except for a syntactic convenience - showing only *some* of the existing overloads is actively harmful. In C# I know how resolution works and I know that the compiler shows methods defined on an object. In Nim, this isn't so, and missing some of the overloads can easily be puzzling (shouldn't there be an overload that applies? I remember I saw it in the documentation... Let me check... Uhm, it's still there, am I even browsing the current version of the docs? Maybe on github docs are more up to date...)

**Comment by Araq** [comment]
Ok, will revert this change then. Have fun with the irrelevant overloads.

**Comment by alehander42** [comment]
@andreaferretti 
no, the point is exactly that the errors will also show you the position in the overloads where the argument doesn't match: so you'll have exactly the same info

`The compiler cannot show with an arrow where the error is` : the arrow will be for each overload, not for the original type. it will be an alternative for the color.

**Comment by alehander42** [comment]
@andreaferretti basically what I meant was that I also prefer the color solution , but if one doesn't have colors, we can use arrows: otherwise I agree with you, we need the same info, just visualized differently

**Comment by timotheecour** [comment]
@Araq @andreaferretti
> Ok, will revert this change then. Have fun with the irrelevant overloads.

I think we can get the best of both worlds, see proposal: [`nim --verboselog:/tmp/log.txt ...` to log more context when needed while not obfuscating stdout · Issue #8789 · nim-lang/Nim](https://github.com/nim-lang/Nim/issues/8789)

**Comment by alehander42** [comment]
Ok, so what is the consensus? @andreaferretti you thumbed up my last comment: does it mean you agree that https://github.com/nim-lang/Nim/issues/8200#issuecomment-402243841 is a good solution?

I still think the current errors use tooooooooo much space and the proposed solutions satisfy all your requirements with a much better UX .

**Comment by alehander42** [comment]
Ok, so what is the consensus? @andreaferretti you thumbed up my last comment: does it mean you agree that #8200 (comment) is a good solution (sorry for the lack of clarity before that)?

**Comment by andreaferretti** [comment]
@alehander42 Yes, I agree that the information can be organized to take less visual space, using colors where available and arrows otherwise (I hope there is a library for that! :-D). My only requirement is not to lose the information that we have available now to free up space - I am ok with every rearranging that preserves the semantics, and I think that colors can be very useful for that

**Comment by krux02** [comment]
I like the error messages as they are. They are a bit verbose, but I am fine with it, and I got used to it. I would be ok to use the nimvm to register an error message callback that takes care to format error messages to arbitrarily. if somebody then develops objectively better error messages, I am ok to take them as the new default. But I don't like the discussions about opinions how error messages should look like.

The suggested error message is more compact, yes. But it is not clear what the {} means to any programmer. It is not a convention that I have seen anywhere.

**Comment by alehander42** [comment]
@krux02 

> I like the error messages as they are.

I also like a lot of the error messages, obviously some of them can be even better with some love.

> and I got used to it

"Getting used" to something is usually a symptom of it being suboptimal :)
We all have different pain points.

> I would be ok to use the nimvm to register an error message callback that takes care to format error messages to arbitrarily.

That's a very cool idea, but I am not sure it's top priority: the more important thing is to have a nice default and to be equally useful to pro users and beginners in the same time.

>  if somebody then develops objectively better error messages .. But I don't like the discussions about opinions how error messages should look like.

People proposed several good designs: there is no need to dismiss them.

Objectively some of the proposed solutions preserve all of the pro-s of the existing solution,  + a better grouped output/way easier to read quickly output. 

The `{}` doesn't seem optimal: but the discussion in the later comments is about https://github.com/nim-lang/Nim/issues/8200#issuecomment-402243841 .

It doesn't use `{}`, it uses colors or `---^` markers which are both obvious. For no-color mode, one can use something different than `{}`.

**Comment by alehander42** [comment]
@krux02 

> But I don't like the discussions about opinions how error messages should look like.

Error messages are the main interface of the compiler: they're literally 90% of the whole interaction of the user with it. It's extremely important how they look

**Comment by krux02** [comment]
@alehander42 well that is a fair point.
